src/jemalloc.c

```diff
@@ -2384,26 +2384,19 @@ imalloc_sample(static_opts_t *sopts, dynamic_opts_t *dopts, tsd_t *tsd,
     size_t usize, szind_t ind) {
 	void *ret;
 
-	/*
-	 * For small allocations, sampling bumps the usize.  If so, we allocate
-	 * from the ind_large bucket.
-	 */
-	szind_t ind_large;
-
 	dopts->alignment = prof_sample_align(dopts->alignment);
 	if (usize <= SC_SMALL_MAXCLASS) {
-		assert(((dopts->alignment == 0) ?
-		    sz_s2u(SC_LARGE_MINCLASS) :
-		    sz_sa2u(SC_LARGE_MINCLASS, dopts->alignment))
-			== SC_LARGE_MINCLASS);
-		ind_large = sz_size2index(SC_LARGE_MINCLASS);
-		size_t bumped_usize = sz_s2u(SC_LARGE_MINCLASS);
-		ret = imalloc_no_sample(sopts, dopts, tsd, bumped_usize,
-		    bumped_usize, ind_large);
+		size_t bumped_usize = PAGE_CEILING(usize);
+		arena_t *arena;
+		if (arena_get_from_ind(tsd, dopts->arena_ind, &arena)) {
+			return NULL;
+		}
+		ret = arena_large_palloc(tsd_tsdn(tsd), arena, bumped_usize, dopts->alignment,
```

Comment:
We should still go through `imalloc_no_sample` in order to handle all the requirements, e.g. if there's a specific tcache id asked. Also it's helpful to keep the high level flow same for small and large sizes, with only small being bumped up a bit. If we update the usize using sa2u above, imalloc_no_sample should work correctly?

Reply:
I just realized that `imalloc_no_sample` cannot be used because it may go through the small size class route and allocate part of a slab to us. So it appears we would have to short cut a bit here. Let me look more into it.